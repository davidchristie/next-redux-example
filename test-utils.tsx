import React from "react";
import { Provider } from "react-redux";
import { createStore, Dispatch, Store, Action } from "redux";
import { render } from "@testing-library/react";
import { exampleInitialState, RootState } from "./reducer";

type Context = ReturnType<typeof renderConnectedComponent>;

type GetContext = () => Context;

const createWrapper = (initialState: RootState) => {
  const reducer = (state = initialState) => state;
  const store = createStore(reducer);
  const dispatch = mockDispatch(store);
  const Wrapper: React.FunctionComponent = ({ children }) => (
    <Provider store={store}>{children}</Provider>
  );
  return { dispatch, Wrapper };
};

const mockDispatch = (store: Store): jest.Mock<Dispatch> => {
  const dispatch = jest.fn(store.dispatch);
  store.dispatch = dispatch;
  return dispatch;
};

const renderConnectedComponent = (
  element: React.ReactElement,
  { initialState = exampleInitialState } = {}
) => {
  const { dispatch, Wrapper } = createWrapper(initialState);
  const result = render(element, {
    wrapper: Wrapper
  });
  return { dispatch, result };
};

export const describeComponent = <P extends {}>(
  Component: React.ComponentType<P>,
  props: P,
  fn: (getContext: GetContext) => void
) => {
  describe(`${Component.displayName} Component`, () => {
    let context: Context;

    beforeEach(() => {
      context = renderConnectedComponent(<Component {...props} />);
    });

    it("Has display name", () => {
      expect(Component.displayName).toBeDefined();
    });

    fn(() => context);
  });
};

export const describeComponentWithReduxState = <P extends {}>(
  Component: React.ComponentType<P>,
  props: P,
  state: Partial<RootState>,
  fn: (getContext: GetContext) => void
) => {
  describe(`${Component.displayName} with Redux state: ${JSON.stringify(
    state
  )}`, () => {
    let context: Context;

    beforeEach(() => {
      context = renderConnectedComponent(<Component {...props} />, {
        initialState: { ...exampleInitialState, ...state }
      });
    });

    fn(() => context);
  });
};

export const describeWhenButtonIsClicked = (
  getContext: GetContext,
  text: string,
  fn: (getContext: GetContext) => void
) => {
  describe(`When "${text}" button is clicked`, () => {
    let context: Context;

    beforeEach(() => {
      context = getContext();
      const button = context.result.getByText(text, {
        selector: "button"
      });
      button.click();
    });

    fn(() => context);
  });
};

export const itHasLink = (
  getContext: GetContext,
  text: string,
  href: string
) => {
  it(`Has link "${text}" to "${href}"`, () => {
    const { result } = getContext();
    expect(
      result.getByText(text, {
        selector: `a[href="${href}"]`
      })
    ).not.toBeNull();
  });
};

export const itShouldDispatchAction = (
  getContext: GetContext,
  action: Action
) => {
  it(`Dispatches "${action.type}" action`, () => {
    const { dispatch } = getContext();
    expect(dispatch).toBeCalledWith(action);
  });
};

export const itShouldMatchSnapshot = (getContext: GetContext) => {
  it("Should match snapshot", () => {
    const { result } = getContext();
    expect(result.asFragment()).toMatchSnapshot();
  });
};

export const itShowsText = (getContext: GetContext, text: string) => {
  it(`Shows "${text}"`, () => {
    const { result } = getContext();
    expect(result.getByText(text)).not.toBeNull();
  });
};
