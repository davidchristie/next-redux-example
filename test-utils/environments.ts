const environments = ["development", "production", "test"];

export const describeAllEnvironments = (fn: (environment: string) => void) => {
  const originalEnv = process.env;

  beforeEach(() => {
    jest.resetModules();
    const mockEnv = { ...originalEnv };
    delete mockEnv.NODE_ENV;
    process.env = mockEnv;
  });

  afterEach(() => {
    process.env = originalEnv;
  });

  describe.each(environments.map(environment => [environment]))(
    'In Environment: "%s"',
    environment => {
      (process.env as any).NODE_ENV = environment;

      fn(environment);
    }
  );
};
