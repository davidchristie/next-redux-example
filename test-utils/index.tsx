export * from "./components";
export * from "./environments";
export * from "./objects";
export * from "./reducers";
export * from "./stores";
