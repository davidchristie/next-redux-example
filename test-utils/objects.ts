export const itShouldHaveMethod = <T>(instance: T, method: keyof T) => {
  it(`Should have ${method} method`, () => {
    expect(instance[method]).toBeInstanceOf(Function);
  });
};
