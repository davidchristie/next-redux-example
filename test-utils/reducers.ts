import { Reducer, AnyAction } from "redux";

export const itShouldHaveNextState = <S extends {}>(
  name: string,
  {
    action,
    expectedNextState,
    previousState,
    reducer
  }: {
    action: AnyAction;
    expectedNextState: S;
    previousState: S;
    reducer: Reducer<S>;
  }
) => {
  it(name, () => {
    const actualNextState = reducer(previousState, action);
    expect(actualNextState).toEqual(expectedNextState);
  });
};
