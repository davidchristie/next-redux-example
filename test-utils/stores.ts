import { Dispatch, Store } from "redux";

export const mockDispatch = (store: Store): jest.Mock<Dispatch> => {
  const dispatch = jest.fn(store.dispatch);
  store.dispatch = dispatch;
  return dispatch;
};
