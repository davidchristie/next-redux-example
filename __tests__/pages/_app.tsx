import { AppProps } from "next/app";
import { Router } from "next/dist/client/router";
import App, { AppContent } from "../../pages/_app";
import { describeComponent, itShouldMatchSnapshot } from "../../test-utils";
import { NextComponentType } from "next/dist/next-server/lib/utils";

describeComponent<AppProps>(
  App,
  { Component: () => null, pageProps: {}, router: {} as Router },
  getContext => {
    itShouldMatchSnapshot(getContext);

    describe("Initial Props", () => {
      it("Should match snapshot", async () => {
        const Component: NextComponentType = () => null;
        const initialProps = await AppContent.getInitialProps!({
          AppTree: () => null,
          Component,
          ctx: {} as any,
          router: {} as Router
        });
        expect(initialProps).toMatchSnapshot();
      });

      describe("If Component has getInitialProps static method", () => {
        it("Should match snapshot", async () => {
          const Component: NextComponentType = () => null;
          Component.getInitialProps = () => Promise.resolve({});
          const initialProps = await AppContent.getInitialProps!({
            AppTree: () => null,
            Component,
            ctx: {} as any,
            router: {} as Router
          });
          expect(initialProps).toMatchSnapshot();
        });
      });
    });
  }
);
