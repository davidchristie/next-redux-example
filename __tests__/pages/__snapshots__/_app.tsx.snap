// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component Initial Props If Component has getInitialProps static method Should match snapshot 1`] = `
Object {
  "pageProps": Object {},
}
`;

exports[`App Component Initial Props Should match snapshot 1`] = `
Object {
  "pageProps": Object {},
}
`;

exports[`App Component Should match snapshot 1`] = `<DocumentFragment />`;
